lib/plugins/aws/deploy/lib/ensure-valid-bucket-exists.js

```diff
@@ -0,0 +1,147 @@
+      let result;
+      try {
+        result = await this.provider.request('S3', 'getBucketLocation', {
+          Bucket: this.bucketName,
+        });
+      } catch (err) {
+        throw new ServerlessError(
+          `Could not locate deployment bucket. Error: ${err.message}`,
+          'DEPLOYMENT_BUCKET_NOT_FOUND'
+        );
+      }
+
+      if (result.LocationConstraint === '') result.LocationConstraint = 'us-east-1';
+      if (result.LocationConstraint === 'EU') result.LocationConstraint = 'eu-west-1';
+      if (result.LocationConstraint !== this.provider.getRegion()) {
+        throw new ServerlessError(
+          'Deployment bucket is not in the same region as the lambda function',
+          'DEPLOYMENT_BUCKET_INVALID_REGION'
+        );
+      }
+      // If above is satisfied, then custom S3 bucket is valid
+      return;
+    }
+
+    // Ensure to set bucket name if it can be resolved
+    // Result of this operation will determine how further validation will be performed
+    try {
+      await this.setBucketName();
+    } catch (err) {
+      // If there is a validation error with expected message, it means that logical resource for
+      // S3 bucket does not exist and we want to proceed with handling that situation
+      if (
+        err.providerError.code !== 'ValidationError' ||
+        !err.message.includes('does not exist for stack')
+      ) {
+        throw err;
+      }
+    }
+
+    // If bucket name is set, it means it's defined as a part of CloudFormation template (custom bucket case was handled by logic above)
+    if (this.bucketName) {
+      if (!(await this.checkIfBucketExists(this.bucketName))) {
+        // It means that bucket was removed manually but is still a part of the CloudFormation stack, we cannot manually fix it
+        throw new ServerlessError(
+          'Deployment bucket has been removed manually. Please recreate it or remove your service and attempt to deploy it again',
+          'DEPLOYMENT_BUCKET_REMOVED_MANUALLY'
+        );
+      }
+      return;
+    }
+
+    legacy.log(
+      'Deployment bucket not found. Updating stack to include deployment bucket definition.'
+    );
+    log.info(
+      'Deployment bucket not found. Updating stack to include deployment bucket definition.'
+    );
+    const stackName = this.provider.naming.getStackName();
+
+    // This is situation where the bucket is not defined in the template at all
+    // It covers the case where someone was using custom deployment bucket
+    // but removed that setting from the configuration
+    const getTemplateResult = await this.provider.request('CloudFormation', 'getTemplate', {
+      StackName: stackName,
+      TemplateStage: 'Original',
+    });
+
+    const templateBody = _.merge(
+      JSON.parse(getTemplateResult.TemplateBody),
+      this.serverless.service.provider.coreCloudFormationTemplate
+    );
+    let stackTags = { STAGE: this.provider.getStage() };
```

Comment:
I've decided to not abstract setup of all these properties used in `updateStack` as a part of this PR as I already have it simplified in WIP branch for changesets, I will address that change there